src/main/scala/cromwell/engine/backend/runtimeattributes/CromwellRuntimeAttributes.scala

```diff
@@ -24,156 +27,159 @@ case class CromwellRuntimeAttributes(docker: Option[String],
 object CromwellRuntimeAttributes {
   private val log = LoggerFactory.getLogger("RuntimeAttributes")
 
-  def apply(wdlRuntimeAttributes: RuntimeAttributes, backendType: BackendType): CromwellRuntimeAttributes = {
-    val attributeMap = AttributeMap(wdlRuntimeAttributes.attrs)
+  def apply(wdlRuntimeAttributes: RuntimeAttributes, backendCall: BackendCall): Try[CromwellRuntimeAttributes] = {
+    for {
+      keys <- validateKeys(wdlRuntimeAttributes, backendCall.backend.backendType)
+      attributes <- TryUtil.sequenceMap(wdlRuntimeAttributes.evaluate(backendCall.lookupFunction(backendCall.locallyQualifiedInputs), backendCall.engineFunctions))
+      validatedAttributes <- validateRuntimeAttributes(attributes)
+    } yield validatedAttributes
+  }
+
+  def validateKeys(wdlRuntimeAttributes: RuntimeAttributes, backendType: BackendType): Try[RuntimeAttributes] = {
     /**
-     *  Warn if keys are found that are unsupported on this backend.  This is not necessarily an error, at this point
-     *  the keys are known to be supported on at least one backend other than this.
-     */
-    attributeMap.unsupportedKeys(backendType) foreach log.warn
-
-    val attributeMapNel = validateAttributeMap(attributeMap, backendType)
-    val runtimeAttributeNel = validateRuntimeAttributes(attributeMap)
-    val validatedRuntimeAttributes = (attributeMapNel |@| runtimeAttributeNel) { (_, r) => r }
-
-    validatedRuntimeAttributes match {
-      case Success(r) => r
-      case Failure(f) =>
-        throw new IllegalArgumentException() with ThrowableWithErrors {
-          val message = "RuntimeAttribute is not valid."
-          val errors = f
-        }
+      *  Warn if keys are found that are unsupported on this backend.  This is not necessarily an error, at this point
+      *  the keys are known to be supported on at least one backend other than this.
+      */
+    unsupportedKeys(wdlRuntimeAttributes.attrs.keys, backendType) foreach log.warn
+    val requiredKeysNel = validateRequiredKeys(wdlRuntimeAttributes.attrs, backendType)
+    val unknownKeysNel = validateUnknownKeys(wdlRuntimeAttributes.attrs, backendType)
+    (requiredKeysNel |@| unknownKeysNel) { (_, _) } match {
+      case scalaz.Success(x) => scala.util.Success(wdlRuntimeAttributes)
+      case scalaz.Failure(e) => scala.util.Failure(new IllegalArgumentException(e.list.mkString("\n")))
     }
   }
 
+  def unsupportedKeys(keys: Iterable[String], backendType: BackendType): Seq[String] = {
+    val supportedKeys = backendType.supportedKeys map { _.key }
+    val unsupportedKeys = keys.toSet -- supportedKeys
+
+    if (unsupportedKeys.isEmpty) Vector.empty
+    else Vector(s"Found unsupported keys for backend '$backendType': " + unsupportedKeys.toSeq.sorted.mkString(", "))
```

Comment:
It seems weird to have a method with a `Seq[String]` return type which never returns a Seq with more than one element. Why not just return `unsupportedKeys.toSeq.sorted` in this case ?